### DISCUSSION TITLE
unsafe for/while

### DISCUSSION THREAD

**Discussion by LogicalError** [dotnet/csharplang#6845] [Language Ideas]
When a loop cannot be *trivially* determined to halt, the compiler would refuse to compile it unless you put an "unsafe" in front of it (or some other keyword)
* This makes it clear when a loop could potentially be an infinite loop, reducing the loops you would need to check when you do encounter one
* Lets the compiler help you detect accidental infinite for loops where the wrong variable is iterated on, or because of signed/unsigned mismatch + overflow of the signed variable

**Comment by svick** [comment]
I don't see this being included in the compiler by default (certainly not as an error, and probably not even as a warning), since most people would see it as unnecessary ceremony.

But it could make sense as an optional warning, one you could even write yourself as an analyzer. Though this means you wouldn't get a dedicated syntax for this and would have to repurpose some existing syntax, e.g. a `// unsafe` comment.

**Comment by CyrusNajmabadi** [comment]
Yup, an analyzer would be best here.
